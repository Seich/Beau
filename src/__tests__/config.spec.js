const yaml = require('js-yaml')
const Config = require('../config')

const requireg = require('requireg')
requireg.resolving = false

describe('Config', () => {
    it('should load valid config keys', () => {
        const doc = yaml.load(`
            version: 1
            endpoint: http://martianwabbit.com
            shouldntBeAdded: true
        `)

        const config = new Config(doc)
        expect(config.ENDPOINT).toBe(doc.endpoint)
        expect(config.VERSION).toBe(doc.version)
        expect(config.shouldntBeAdded).toBeUndefined()
    })

    it('should load requests', () => {
        const doc = yaml.load(`
            endpoint: http://example.com

            GET /profile: get-profile
            GET /posts: get-posts
            GET /posts/1: get-post
            GET /user:
                alias: user
                headers:
                    hello: world
        `)

        const config = new Config(doc)
        expect(Object.keys(config.REQUESTS).length).toBe(4)
    })

    it('should set up defaults for all requests', () => {
        const doc = yaml.load(`
            version: 1
            endpoint: 'http://example.com'

            defaults:
                HEADERS:
                    authentication: hello

            GET /posts/1: get-post
            GET /user:
                alias: user
                headers:
                    hello: world
        `)

        const config = new Config(doc)

        expect(config).toMatchSnapshot()
        Object.values(config.REQUESTS).forEach((r) => {
            expect(r.HEADERS.authentication).toMatch('hello')
        })
    })

    it('should load multiple hosts', () => {
        const doc = yaml.load(`
          version: 1
          endpoint: http://example.org

          defaults:
            HEADERS:
              hello: mars

          GET /e1: e1

          hosts:
            - host: com
              endpoint: http://example.com

              defaults:
                HEADERS:
                  hello: world
                  world: hello

              GET /e2: e2
              GET /posts: posts

            - host: net
              endpoint: http://example.net

              defaults:
                HEADERS:
                  hello: world
                  world: bye

              GET /e3: e3
              GET /posts: posts

            - host: info
              endpoint: http://example.info

              GET /posts: posts
        `)

        let config = new Config(doc)

        expect(config).toMatchSnapshot()
    })

    it('should namespace all aliases within an host', () => {
        const doc = yaml.load(`
            hosts:
              - host: test1
                endpoint: http://example.com
                GET /posts: posts
              - host: test2
                endpoint: http://example.net
                GET /posts: posts
        `)

        let config = new Config(doc)

        expect(config.REQUESTS[0].ALIAS).toBe('test1:posts')
        expect(config.REQUESTS[1].ALIAS).toBe('test2:posts')
    })

    it(`should throw if host doesn't have a host key`, () => {
        const doc = yaml.load(`
            hosts:
              - endpoint: http://example.com
                GET /posts: posts

              - host: test2
                endpoint: http://example.net
                GET /posts: posts
        `)

        expect(() => new Config(doc)).toThrow()
    })

    it(`should merge host settings with global settings`, () => {
        const doc = yaml.load(`
            defaults:
              headers:
                hello: 1

            hosts:
              - host: test
                endpoint: http://example.net
                GET /posts: posts

              - host: test2
                endpoint: http://example.org
                defaults:
                  headers: false

                GET /posts: posts
        `)

        let config = new Config(doc)
        expect(config.REQUESTS[0].HEADERS.hello).toBe(1)
    })

    it(`should allow different settings for the same request`, () => {
        const doc = yaml.load(`
      host: https://example.com
      GET /1:
        - alias: req1
          headers:
            request: 1
        - alias: req2
          headers:
            request: 2
    `)

        let config = new Config(doc)
        expect(config.REQUESTS.length).toBe(2)
    })
})
