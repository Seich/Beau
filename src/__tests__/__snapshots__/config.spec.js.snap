// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config should load multiple hosts 1`] = `
Config {
  "CACHE": false,
  "DEFAULTS": Object {
    "HEADERS": Object {
      "hello": "mars",
    },
  },
  "ENDPOINT": "http://example.org",
  "ENVIRONMENT": Object {},
  "HOSTS": Array [
    Object {
      "GET /e2": "e2",
      "GET /posts": "posts",
      "defaults": Object {
        "HEADERS": Object {
          "hello": "world",
          "world": "hello",
        },
      },
      "endpoint": "http://example.com",
      "host": "com",
    },
    Object {
      "GET /e3": "e3",
      "GET /posts": "posts",
      "defaults": Object {
        "HEADERS": Object {
          "hello": "world",
          "world": "bye",
        },
      },
      "endpoint": "http://example.net",
      "host": "net",
    },
    Object {
      "GET /posts": "posts",
      "endpoint": "http://example.info",
      "host": "info",
    },
  ],
  "PLUGINS": Array [],
  "VERSION": 1,
  "configKeys": Array [
    "VERSION",
    "CACHE",
    "ENDPOINT",
    "PLUGINS",
    "DEFAULTS",
    "ENVIRONMENT",
    "HOSTS",
  ],
  "defaultConfigValues": Object {
    "CACHE": false,
    "DEFAULTS": Object {},
    "ENDPOINT": "",
    "ENVIRONMENT": Object {},
    "HOSTS": Array [],
    "PLUGINS": Array [],
    "VERSION": 1,
  },
  "doc": Object {
    "GET /e1": "e1",
    "defaults": Object {
      "HEADERS": Object {
        "hello": "mars",
      },
    },
    "endpoint": "http://example.org",
    "hosts": Array [
      Object {
        "GET /e2": "e2",
        "GET /posts": "posts",
        "defaults": Object {
          "HEADERS": Object {
            "hello": "world",
            "world": "hello",
          },
        },
        "endpoint": "http://example.com",
        "host": "com",
      },
      Object {
        "GET /e3": "e3",
        "GET /posts": "posts",
        "defaults": Object {
          "HEADERS": Object {
            "hello": "world",
            "world": "bye",
          },
        },
        "endpoint": "http://example.net",
        "host": "net",
      },
      Object {
        "GET /posts": "posts",
        "endpoint": "http://example.info",
        "host": "info",
      },
    ],
  },
  "requests": Array [
    Object {
      "ALIAS": "e1",
      "ENDPOINT": "http://example.org",
      "HEADERS": Object {
        "hello": "mars",
      },
      "REQUEST": "GET /e1",
    },
    Object {
      "ALIAS": "com:e2",
      "ENDPOINT": "http://example.com",
      "HEADERS": Object {
        "hello": "world",
        "world": "hello",
      },
      "REQUEST": "GET /e2",
    },
    Object {
      "ALIAS": "com:posts",
      "ENDPOINT": "http://example.com",
      "HEADERS": Object {
        "hello": "world",
        "world": "hello",
      },
      "REQUEST": "GET /posts",
    },
    Object {
      "ALIAS": "net:e3",
      "ENDPOINT": "http://example.net",
      "HEADERS": Object {
        "hello": "world",
        "world": "bye",
      },
      "REQUEST": "GET /e3",
    },
    Object {
      "ALIAS": "net:posts",
      "ENDPOINT": "http://example.net",
      "HEADERS": Object {
        "hello": "world",
        "world": "bye",
      },
      "REQUEST": "GET /posts",
    },
    Object {
      "ALIAS": "info:posts",
      "ENDPOINT": "http://example.info",
      "HEADERS": Object {
        "hello": "mars",
      },
      "REQUEST": "GET /posts",
    },
  ],
}
`;

exports[`Config should set up defaults for all requests 1`] = `
Config {
  "CACHE": false,
  "DEFAULTS": Object {
    "HEADERS": Object {
      "authentication": "hello",
    },
  },
  "ENDPOINT": "http://jsonplaceholder.typicode.com",
  "ENVIRONMENT": Object {},
  "HOSTS": Array [],
  "PLUGINS": Array [],
  "VERSION": 1,
  "configKeys": Array [
    "VERSION",
    "CACHE",
    "ENDPOINT",
    "PLUGINS",
    "DEFAULTS",
    "ENVIRONMENT",
    "HOSTS",
  ],
  "defaultConfigValues": Object {
    "CACHE": false,
    "DEFAULTS": Object {},
    "ENDPOINT": "",
    "ENVIRONMENT": Object {},
    "HOSTS": Array [],
    "PLUGINS": Array [],
    "VERSION": 1,
  },
  "doc": Object {
    "GET /posts/1": "get-post",
    "GET /user": Object {
      "alias": "user",
      "headers": Object {
        "hello": "world",
      },
    },
    "defaults": Object {
      "HEADERS": Object {
        "authentication": "hello",
      },
    },
    "endpoint": "http://jsonplaceholder.typicode.com",
    "version": 1,
  },
  "requests": Array [
    Object {
      "ALIAS": "get-post",
      "ENDPOINT": "http://jsonplaceholder.typicode.com",
      "HEADERS": Object {
        "authentication": "hello",
      },
      "REQUEST": "GET /posts/1",
    },
    Object {
      "ALIAS": "user",
      "ENDPOINT": "http://jsonplaceholder.typicode.com",
      "HEADERS": Object {
        "authentication": "hello",
        "hello": "world",
      },
      "REQUEST": "GET /user",
    },
  ],
}
`;
